# Advent of Code 2023 Day 1: Trebuchet?!

https://adventofcode.com/2023/day/1

Something is wrong with global snow production, and the elves are launching me into the sky via trebuchet. Its calibration document has been scrambled.

## Part 1

In each line of text, identify the first and last digit and combine these into a two-digit number. Calculate the sum of these values.

In [1]:
sample_input = '''1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet'''

12 + 38 + 15 + 77 = 142

The puzzle input is much longer, we'll need a program to calculate the result for us. Let's work it out with the sample of manageable size. First, we need to be able to deal with one line at a time.

In [2]:
lines = sample_input.split('\n')
for line in lines:
    print(line)

1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet


I used the regular expression module to search for all the digits in each line.

In [3]:
import re

digits = re.findall('[0-9]', lines[1])

digits

['3', '8']

In [4]:
last_digit = digits[-1]

last_digit

'8'

We will concatentate these digits represented as strings and then convert to an integer, and keep a running total as we work our way through all the lines.

In [5]:
def decode_part1(humongous_string):
    lines = humongous_string.split('\n')
    sum = 0
    for line in lines:
        digits = re.findall('[0-9]', line)
        calibration_value = int(digits[0] + digits[-1])
        sum = sum + calibration_value
    return sum
    

In [6]:
decode_part1(sample_input)

142

OK, we have a working program. Now we get the puzzle input.

In [ ]:
with open('data/day01_in.txt', 'r') as f:
    puzzle_input = f.read()

print(puzzle_input)

In [8]:
len(puzzle_input)

21936

The puzzle input has 21936 lines. I told you we'd need a program!

In [9]:
decode_part1(puzzle_input)

53651

## Part two

Now we also need to identify numbers that are spelled out with letters. 

In [10]:
sample_input = '''two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen'''

The calibration values are 29, 83, 13, 24, 42, 14, and 76, with a sum of 281.

I started by using string replace to overwrite words with numerals. But that didn't work because some words overlap, like eightwo. Next I tried the following:

In [11]:
def digit(s):
    if s.find('0') >=0 or s.lower().find('zero') >=0:
        return '0'
    if s.find('1') >=0 or s.lower().find('one') >=0:
        return '1'
    if s.find('2') >=0 or s.lower().find('two') >=0:
        return '2'
    if s.find('3') >=0 or s.lower().find('three') >=0:
        return '3'
    if s.find('4') >=0 or s.lower().find('four') >=0:
        return '4'
    if s.find('5') >=0 or s.lower().find('five') >=0:
        return '5'
    if s.find('6') >=0 or s.lower().find('six') >=0:
        return '6'
    if s.find('7') >=0 or s.lower().find('seven') >=0:
        return '7'
    if s.find('8') >=0 or s.lower().find('eight') >=0:
        return '8'
    if s.find('9') >=0 or s.lower().find('nine') >=0:
        return '9'
    return ''

digit('eightwothree')

'2'

This approach still does not work with some lines, since it doesn't find numbers in order from left to right or right to left. We can use this function to recognize spelled-out digits, but we need start with the first and last characters of each line and work our way in until we have a digit within the substring.

In [12]:
def get_first(s):
    first_digit = ''
    num_chars = 1
    while (first_digit == '') & (num_chars <= len(s)):
        substr = s[:num_chars]
        first_digit = digit(substr)
        num_chars = num_chars + 1
    return first_digit

get_first('eightwothree')

'8'

In [13]:
def get_last(s):
    last_digit = ''
    num_chars = 1
    while (last_digit == '') & (num_chars <= len(s)):
        substr = s[-1*num_chars:]
        last_digit = digit(substr)
        num_chars = num_chars + 1
    return last_digit

get_last('eightwothree')

'3'

In [14]:
def decode_part2(humongous_string):
    lines = humongous_string.split('\n')
    sum = 0
    for line in lines:
        calibration_value = int(get_first(line) + get_last(line))
        sum = sum + calibration_value
    
    return sum

In [15]:
decode_part2(sample_input)

281

In [16]:
decode_part2(puzzle_input)

53894